### ISSUE TITLE
[discussion] Integration with OBS

### ISSUE THREAD

**Issue by mudler** [rancher-sandbox/cOS-toolkit#1349]
**Is your feature request related to a problem? Please describe.**
Currently, OBS builds and os2 builds aren't strictly tight, and this is becoming more an issue in both releasing and iterating in development, because of "historical" reasons

Current state:
- https://github.com/rancher-sandbox/rancher-node-image builds from the same image as https://github.com/rancher-sandbox/os2, re-applying framework files (both os2, and cOS served ones) to build a custom ISO with a YAML config file, in order to be used for bootstrapping clusters with custom configurations
- https://github.com/rancher-sandbox/os2 contains framework files (cOS + os2 specific) that needs to be shipped with the CLI in order for the system to function properly
- https://github.com/rancher-sandbox/os2 is based on the SLE Micro for Rancher images, which contains already the elemental CLI and ros-installer. So at this point we have no mapping already between framework files <-> CLI/ros-installer. as we just consume them from the base image. This is fine - as long as we keep using that repository as "overlay" to customize the base image.

Here is my proposal here ( to refine still ) :

- rancher-node-image repo shouldn't exist. Users, in order to build ISO with their custom yaml config don't need to build a docker image, and neither build a container image which embeds framework files. They just need a source docker image, which, in worse case they would base against another one if they need customizations, if any.
- The source should be only os2, built from OBS and our CI (testing, gating breaking changes, and so forth) from tagged releases
  - when iterating, we can change base image to a `-testing`, which contains the latest tagged binaries of our projects.
- at that point, framework is handled in os2, and elemental-cli and ros-installer by OBS, served from the base image.
 
 Changes to the framework needs to be tied to the CLI and ros-installer, so development iteration becomes a HELL (but, due to OBS limitations I can't do nothing about it):
   - testing local changes requires - local build of CLI, ros-installer and framework files. This we can probably automate with some make targets to make our life easier, or by any other means
   - Consideration: PRs won't represent changes at hand anymore. For example, we have testing images with CLI at version X - while released is at Y - so requires any changes priorly to the changeset to land in the respective repositories first.
Suggestion: github action and tests to run against test-images. while instead OBS runs builds against stable-images containing the tools at a pinned set
     
# Releases:
   - When we tag os2, framework files and latest tags of respective projects are aligned and be properly tagged into the base image.
   - Bump CLI and ros-installer to tags used from testing images at the os2 tag
   - Build elemental-teal image from os2 sources against the image which contains tool at tagged versions

**Comment by davidcassany** [comment]
I agree the main source should be os2 repo so the `rancher-node-image` is not needed.

It is still a bit unclear to me what gets built in OBS and what not, but IMHO we could achieve a pretty decent automation and link between our repos and OBS.

The OS2 image should be built in OBS with, at least, two flavors: `tagged` and `latest`. Assume `tagged` is the lastest tag in os2 repo and `latest` is the master/main branch of os2 repo.

An equivalent criteria should be set for elemental (and all other tools we want to build and package in OBS) and make os2 OBS builds to consume those respectively. 

That means we could always have an os2 build from latest commits and one including the latest tag of each repo happening in OBS in an automated fashion. It think this could be a great compromise as a start, such an automation is not hard to configure.

Then taking it even a bit further, we could even consider a third build in OBS, a staging build, where the builds are comming from pinned commits of the repositories (elemental, ros-installer, os2, ...), then we could use this staging builds to prepare next round of repo tagging without blocking commits in master/main branches.

**Comment by kkaempf** [comment]
`rancher-node-image` was created to provide a simple test-bed for SLE based deployments. I agree that it shouldn't exist, but os2's `ros-image-build` didn't work for me reliably.

Everything in OBS is a "test balloon", to better understand the building blocks and find ways to get it accepted into IBS.

`rancher-node-image/Dockerfile` was created because some artefacts (like `meta/cos-light`) aren't RPM-packaged yet.

For the final 'product', the OS image must come from registry.suse.com. That means built in IBS. That means all pieces be available offline. (Could be a meta-RPM with lots of tarballs).

**Comment by kkaempf** [comment]
There are also two (maybe three) goals for image building

- building for Elemental testing, CI, etc.
- building for productisation, with SLE, in the build service

and possibly

- building for Elemental community, using openSUSE, Fedora, etc

**Comment by mudler** [comment]
> `rancher-node-image` was created to provide a simple test-bed for SLE based deployments. I agree that it shouldn't exist, but os2's `ros-image-build` didn't work for me reliably.

Then we should fix `ros-image-build` :)

> 
> Everything in OBS is a "test balloon", to better understand the building blocks and find ways to get it accepted into IBS.
> 
> `rancher-node-image/Dockerfile` was created because some artefacts (like `meta/cos-light`) aren't RPM-packaged yet.
> 
> For the final 'product', the OS image must come from registry.suse.com. That means built in IBS. That means all pieces be available offline. (Could be a meta-RPM with lots of tarballs).

That is indeed what os2 recent changes are about - now can be just built now from a pristine base image which contains only the compiled binaries. everything else is provided by that repository and can be used as a final-stage during build.

**Comment by mudler** [comment]
Relevant cards from the discussion have been created, posting here the action items notes for the records:

- Rename Dockerfile to Dockerfile.e2e
- Explore SCM integration with OBS automatically with service file :
  - get Dockerfile.obs (  https://build.opensuse.org/package/view_file/home:kwk:elemental-testing:images/elemental-node-image/Dockerfile?expand=1  ) from elemental repository 
  - hook master/devel and tags to the non-testing repository 
- Drop rancher-node-image 
- Klaus for top level Elemental in OBS/IBS
  - naming on OBS/IBS projects (?)
- elemental teal master tests should differ from PRs to run against OBS-devel built packages.
  Dockerfile.e2e should point to the testing images, built against master while running tests in PR (card to be created)
  While building on master, it should instead point of packages built from OBS with the framework file pointing to rancher/elemental's version
- We should have a mechanism to print out packages coming from the base image as part of the testing/release process in rancher/elemental 


Meeting Notes: 
 
Components status:
- [Image](https://build.opensuse.org/package/show/home:kwk:elemental-testing:images/elemental-operator-image) Installed via helm chart. This is the operator running in the management cluster which talks to the node. 
- [this image ](https://build.opensuse.org/package/show/home:kwk:elemental-testing:images/elemental-builder-image) have qemu and xoriso in order to build qemu and iso files: 
- this [Image](https://build.opensuse.org/package/show/home:kwk:elemental-testing:images/elemental-node-image) is the actual Elemental OS.

Notes:
- There are few packages that should go into sle-micro-rancher, but currently are in the image
- We should rename main [Dockerfile](https://github.com/rancher/elemental/blob/master/Dockerfile) in elemental repo as Dockerfile.e2e as it is for testing purposes only
- Explore OBS SCM Github integration: https://build.opensuse.org/package/view_file/home:kwk:elemental-testing:images/elemental-node-image/Dockerfile?expand=1 so we pick up the Dockerfile in obs from elemental source code instead
- https://github.com/rancher/elemental/blob/master/Dockerfile#L5 - those maps to the elemental-toolkit `meta/cos-light` package in OBS.
- elemental master tests should run against OBS-devel built packages
